### PR TITLE
Homepage posts feed is working

### PR DESCRIPTION
**Pull request by tiakindele**
On his/her home page, the user can now see:
- their post, 
- any public post, 
- as well as posts by friends _where privacy is set to 3 or 4,_
- and posts by friend of friends _where privacy is set to 4._

**we do still need to figure out how to select the specific friend we want to share posts with**

### CODE REVIEW

File: backend/project404_t8/API/viewsets.py [hunk 1]
```diff
@@ -266,21 +240,23 @@ def profileView(request, username):
...
+        (SELECT f2.friend_a_id AS fofid \
+            FROM API_friendship f \
+            JOIN API_friendship f2 ON f.friend_a_id = f2.friend_b_id \
+            WHERE fofid NOT IN (SELECT friend_a_ID FROM API_friendship  \
+            WHERE friend_b_id = %s) AND f.friend_b_id = %s AND fofid != %s) AND privacy_setting = 4 \
+        UNION \
+            SELECT id FROM API_post WHERE (author_id in  \
+            (WITH friends(fid) AS (SELECT friend_b_id FROM API_friendship WHERE friend_a_id=%s) \
+            SELECT * FROM friends WHERE fid != %s GROUP BY fid)  \
+            AND privacy_setting = 3 OR privacy_setting = 4) OR author_id = %s OR  privacy_setting = 6) \
```
Comment (peterweckend) [inline backend/project404_t8/API/viewsets.py]: should there be some ()s somewhere in the "... AND privacy_setting = 3 OR privacy_setting = 4)" part? To tell whether the AND or the OR has precedence?

File: backend/project404_t8/API/viewsets.py [hunk 2]
```diff
@@ -266,21 +240,23 @@ def profileView(request, username):
...
+            SELECT * FROM API_post WHERE id in posts', [int(uid)]*6)
+    except:
+        post = Post.objects.all()
```
Comment (peterweckend) [inline backend/project404_t8/API/viewsets.py]: is this in the case where there's some sort of error with the sqlite query? just out of curiosity

**Comment by bhlewka** [comment]
Oh I didn't know you could do raw SQLite, that's pretty useful

Sent from my iPhone

On Mar 15, 2019, at 11:09 AM, Peter Weckend <notifications@github.com<mailto:notifications@github.com>> wrote:


@peterweckend commented on this pull request.

________________________________

In backend/project404_t8/API/viewsets.py<https://github.com/peterweckend/group-project-cmput404/pull/33#discussion_r266072135>:

> +def postsListView(request):
+    try:
+        uname = request.user
+        uid = uname.id
+        post = Post.objects.raw(' \
+        WITH posts AS (SELECT id FROM API_post WHERE author_id in  \
+        (SELECT f2.friend_a_id AS fofid \
+            FROM API_friendship f \
+            JOIN API_friendship f2 ON f.friend_a_id = f2.friend_b_id \
+            WHERE fofid NOT IN (SELECT friend_a_ID FROM API_friendship  \
+            WHERE friend_b_id = %s) AND f.friend_b_id = %s AND fofid != %s) AND privacy_setting = 4 \
+        UNION \
+            SELECT id FROM API_post WHERE (author_id in  \
+            (WITH friends(fid) AS (SELECT friend_b_id FROM API_friendship WHERE friend_a_id=%s) \
+            SELECT * FROM friends WHERE fid != %s GROUP BY fid)  \
+            AND privacy_setting = 3 OR privacy_setting = 4) OR author_id = %s OR  privacy_setting = 6) \


should there be some ()s somewhere in the "... AND privacy_setting = 3 OR privacy_setting = 4)" part? To tell whether the AND or the OR has precedence?

File: backend/project404_t8/API/viewsets.py (hunk 2 above)
Comment (tiakindele) [reply backend/project404_t8/API/viewsets.py]: it's for the logout-anonymous user error. this makes it safe to logout and still request this list but get nothing back when you lick logout.

File: backend/project404_t8/API/viewsets.py (hunk 1 above)
Comment (tiakindele) [reply backend/project404_t8/API/viewsets.py]: you are absolutely correct sir. fixing that right now.

**Comment by cjlee1** [comment]
Using SQL will help us proceed with everything faster

**Comment by bhlewka** [comment]
looks good

**Pull request merged**